fixtures/parallelizer/__init__.py

```diff
@@ -191,203 +176,172 @@ def __init__(self, config):
+                    num_failed_tests = len(failed_tests)
                     self.sent_tests -= num_failed_tests
                     msg += ' and redistributing {} tests'.format(num_failed_tests)
-                    with SlaveDict.lock:
-                        self.failed_slave_test_groups.append(self.slave_tests.pop(slaveid))
+                    self.failed_slave_test_groups.append(failed_tests)
                 self.print_message(msg, purple=True)
 
-        # Make sure we have a slave for every slave_url
-        for slaveid in list(self.slave_urls):
-            if slaveid not in self.slaves:
-                self._start_slave(slaveid)
-
         # If a slave has lost its base_url for any reason, kill that slave
         # Losing a base_url means the associated appliance died :(
-        for slaveid in list(self.slaves):
-            if slaveid not in self.slave_urls:
-                self.print_message("{}'s appliance has died, deactivating slave".format(slaveid))
-                self.interrupt(slaveid)
-
-    def _start_slave(self, slaveid):
-        devnull = open(os.devnull, 'w')
-        try:
-            base_url = self.slave_urls[slaveid]
-        except KeyError:
-            # race condition: slave was removed from slave_urls when something else decided to
-            # start it; in this case slave_urls wins and the slave should not start
-            return
-
-        # worker output redirected to null; useful info comes via messages and logs
-        slave = subprocess.Popen(
-            ['python', remote.__file__, slaveid, base_url, conf.runtime['env']['ts']],
-            stdout=devnull,
-        )
-        self.slaves[slaveid] = slave
-        self.slave_spawn_count += 1
-        at_exit(slave.kill)
+        for slave in list(self.slaves.values()):
+            if slave.url is None:
+                if slave.process is None:
+                    del self.slaves[slave.id]
+                else:
+                    self.print_message(
+                        "{}'s appliance has died, deactivating slave".format(slave.id))
+                    self.interrupt(slave)
+            else:
+                if slave.process is None:
+                    slave.start()
+                    self.slave_spawn_count += 1
 
-    def send(self, slaveid, event_data):
+    def send(self, slave, event_data):
         """Send data to slave.
 
         ``event_data`` will be serialized as JSON, and so must be JSON serializable
 
         """
         event_json = json.dumps(event_data)
-        with zmq_lock:
-            self.sock.send_multipart([slaveid, '', event_json])
+        self.sock.send_multipart([slave.id, '', event_json])
 
     def recv(self):
-        """Return any unproccesed events from the recv queue"""
-        try:
-            with recv_lock:
-                return self._recv_queue.popleft()
-        except IndexError:
+        # poll the zmq socket, populate the recv queue deque with responses
```

Comment:
This comment doesn't seem to match with what the function actually does, it returns a tuple or data for a single event, and doesn't populate a deque